internals/secrethub/service_aws_init.go

```diff
@@ -0,0 +1,95 @@
+
+	"github.com/secrethub/secrethub-go/internals/api"
+	"github.com/secrethub/secrethub-go/internals/errio"
+)
+
+// Errors
+var (
+	ErrInvalidAWSRegion = errMain.Code("invalid_region").Error("invalid AWS region")
+	ErrRoleAlreadyTaken = errMain.Code("role_taken").Error("a service using that IAM role already exists")
+)
+
+// ServiceAWSInitCommand initializes a service for AWS.
+type ServiceAWSInitCommand struct {
+	description string
+	path        api.DirPath
+	kmsKeyID    string
+	role        string
+	region      string
+	permission  api.Permission
+	io          ui.IO
+	newClient   newClientFunc
+}
+
+// NewServiceAWSInitCommand creates a new ServiceAWSInitCommand.
+func NewServiceAWSInitCommand(io ui.IO, newClient newClientFunc) *ServiceAWSInitCommand {
+	return &ServiceAWSInitCommand{
+		io:        io,
+		newClient: newClient,
+	}
+}
+
+// Run initializes an AWS service.
+func (cmd *ServiceAWSInitCommand) Run() error {
+	repo := cmd.path.GetRepoPath()
+
+	client, err := cmd.newClient()
+	if err != nil {
+		return err
+	}
+
+	cfg := aws.NewConfig()
+	if cmd.region != "" {
+		_, ok := endpoints.AwsPartition().Regions()[cmd.region]
+		if !ok {
+			return ErrInvalidAWSRegion
+		}
+		cfg = cfg.WithRegion(cmd.region)
+	}
+
+	service, err := client.Services().AWS().Create(repo.Value(), cmd.description, cmd.kmsKeyID, cmd.role, cfg)
+	if err == api.ErrCredentialAlreadyExists {
+		return ErrRoleAlreadyTaken
+	} else if err != nil {
+		return err
+	}
+
+	if cmd.permission != 0 {
+		_, err = client.AccessRules().Set(cmd.path.Value(), cmd.permission.String(), service.ServiceID)
+		if err != nil {
+			_, delErr := client.Services().Delete(service.ServiceID)
+			if delErr != nil {
+				fmt.Fprintf(cmd.io.Stdout(), "Failed to cleanup after creating an access rule for %s failed. Be sure to manually remove the created service account %s: %s\n", service.ServiceID, service.ServiceID, err)
+				return errio.Error(delErr)
+			}
+
+			return errio.Error(err)
+		}
+	}
+
+	fmt.Fprintf(cmd.io.Stdout(), "The service %s is now reachable through AWS when the role %s is assumed.\n", service.ServiceID, cmd.role)
+
+	return nil
+}
+
+// Register registers the command, arguments and flags on the provided Registerer.
+func (cmd *ServiceAWSInitCommand) Register(r Registerer) {
+	clause := r.Command("init", "Create a new AWS service account attached to a repository.")
+	clause.Flag("path", "The service account is attached to the repository in this path and when used together with --permission, an access rule is created on the directory in this path.").Required().SetValue(&cmd.path)
+	clause.Flag("kms-key-id", "ID of the KMS-key to be used for encrypting the service's account key.").Required().StringVar(&cmd.kmsKeyID)
+	clause.Flag("role", "ARN of the IAM role that should have access to this service account.").Required().StringVar(&cmd.role)
+	clause.Flag("region", "The AWS region that should be used").StringVar(&cmd.region)
```

Comment:
What does it mean when a `region is used`? Does it mean only hosts within that region can call secrethub and assume this service's role?

Reply:
Mainly important for KMS. The KMS-key should be in this region.

Reply:
Then let's say that in the description 😄

Reply:
Changed the help text to `The AWS region that should be used for KMS.`